pkg/core/klinedriver/klinebuilder.go

```diff
@@ -0,0 +1,195 @@
+package klinedriver
+
+import (
+	"time"
+
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+	log "github.com/sirupsen/logrus"
+)
+
+// KLineBuilder builds klines from trades and trades only
+type KLineBuilder struct {
+	symbol string
+
+	// accBucketMap is a map from interval to accumulated kline of that interval
+	lastBucketMap, accBucketMap map[types.Interval]*Bucket
+	// buffer the the klines that are not yet exposed
+	klinesBuffer map[types.Interval][]*types.KLine
+}
+
+// NewKLineBuilder creates a new KLineBuilder.
+// IMPORTANT: The klines built by this builder are in UTC timezone.
+func NewKLineBuilder(symbol string) *KLineBuilder {
+	return &KLineBuilder{
+		symbol:        symbol,
+		lastBucketMap: make(map[types.Interval]*Bucket),
+		accBucketMap:  make(map[types.Interval]*Bucket),
+		klinesBuffer:  make(map[types.Interval][]*types.KLine),
+	}
+}
+
+// AddInterval adds an interval to the KLineBuilder.
+// AddInterval should be called before any operations on the KLineBuilder.
+func (kb *KLineBuilder) AddInterval(interval types.Interval, initTime types.Time) {
```

Comment:
What's the purpose of wrapping resetKLine  here?
Maybe move the implementation of resetKLine into this public function?
I think "AddInterval" better describes the operations made in "resetKLine".

Reply:
What `resetKLine` does is to resetting the KLine of the given interval in the builder.
I add a method `AddInterval` here because it makes more sense to the caller in comparison with `resetKLine`.
Internally, I use `resetKLine` to update the state of the builder when applicable.

However, I think maybe I should rename it as `resetBucket` at the second thought.
How do you think?

Reply:
If it's for the purpose of internal state management, it should be fine.

Reply:
done.